# DS-SF-36 | 10 | Linear Regression, Part 3 | Assignment | Starter Code

In [1]:
import os

import pandas as pd
pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

from sklearn import feature_selection, linear_model

In [133]:
df = pd.read_csv(os.path.join('..', 'datasets', 'dataset-10-credit.csv'))

In [163]:
df.columns

Index([u'Income', u'Rating', u'Cards', u'Age', u'Education', u'Gender',
       u'Student', u'Married', u'Ethnicity', u'Balance', u'Sex'],
      dtype='object')

A description of the dataset is as follows:

- Income (in thousands of dollars)
- Rating: Credit score rating
- Cards: Number of credit cards owned
- Age
- Education: Years of education
- Gender: Male/Female
- Student: Yes/No
- Married: Yes/No
- Ethnicity: African American/Asian/Caucasian
- Balance: Average credit card debt

> ## Question 1.  Let's explore the quantitative variables that affect `Balance`.  From your preliminary analysis, which 2 variables seem to affect `Balance` the most?  Our goal is interpretation; can we use these 2 variables simultaneously?  Why or why not?

In [162]:
df.corr()

Income    Rating     Cards       Age  Education   Married  \
Income     1.000000  0.791378 -0.018273  0.175338  -0.027692  0.035652   
Rating     0.791378  1.000000  0.053239  0.103165  -0.030136  0.036751   
Cards     -0.018273  0.053239  1.000000  0.042948  -0.051084 -0.009695   
Age        0.175338  0.103165  0.042948  1.000000   0.003619 -0.073136   
Education -0.027692 -0.030136 -0.051084  0.003619   1.000000  0.048911   
Married    0.035652  0.036751 -0.009695 -0.073136   0.048911  1.000000   
Balance    0.463656  0.863625  0.086456  0.001835  -0.008062 -0.005673   
Sex        0.010738 -0.008885  0.022658 -0.004015   0.005049 -0.012452   

            Balance       Sex  
Income     0.463656  0.010738  
Rating     0.863625 -0.008885  
Cards      0.086456  0.022658  
Age        0.001835 -0.004015  
Education -0.008062  0.005049  
Married   -0.005673 -0.012452  
Balance    1.000000 -0.021474  
Sex       -0.021474  1.000000

Answer: n of credit cards and credit score. We could use these both at once, I suppose, but they're related, so it's probably a bad idea.

> ## Question 2.  `Ethnicity`, `Gender`, `Married`, and `Student` are categorical variables.  Go ahead and create dummy variables for all of them.

In [135]:
df['Sex'] = pd.get_dummies(df['Gender'], prefix='Sex').iloc[:, 1:]

In [136]:
df['Married'] = pd.get_dummies(df['Married']).iloc[:, 1:]

In [137]:
df

Income  Rating  Cards  Age  Education ...  Student Married  \
0     14.891     283      2   34         11 ...       No       1   
1    106.025     483      3   82         15 ...      Yes       1   
2    104.593     514      4   71         11 ...       No       0   
3    148.924     681      3   36         11 ...       No       0   
4     55.882     357      2   68         16 ...       No       1   
..       ...     ...    ...  ...        ... ...      ...     ...   
395   12.096     307      3   32         13 ...       No       1   
396   13.364     296      5   65         17 ...       No       0   
397   57.872     321      5   67         12 ...       No       1   
398   37.728     192      1   44         13 ...       No       1   
399   18.701     415      5   64          7 ...       No       0   

            Ethnicity Balance  Sex  
0           Caucasian     333    1  
1               Asian     903    0  
2               Asian     580    1  
3               Asian     964    0  
4           Caucasian     331    1  
..                ...     ...  ...  
395         Caucasian     560    1  
396  African American     480    1  
397         Caucasian     138    0  
398         Caucasian       0    1  
399             Asian     966    0  

[400 rows x 11 columns]

In [139]:
pd.get_dummies(df.Ethnicity, prefix='Ethnicity')
#https://www.youtube.com/watch?v=0s_1IsROgDc

Ethnicity_African American  Ethnicity_Asian  Ethnicity_Caucasian
0                             0                0                    1
1                             0                1                    0
2                             0                1                    0
3                             0                1                    0
4                             0                0                    1
..                          ...              ...                  ...
395                           0                0                    1
396                           1                0                    0
397                           0                0                    1
398                           0                0                    1
399                           0                1                    0

[400 rows x 3 columns]

In [138]:
df

Income  Rating  Cards  Age  Education ...  Student Married  \
0     14.891     283      2   34         11 ...       No       1   
1    106.025     483      3   82         15 ...      Yes       1   
2    104.593     514      4   71         11 ...       No       0   
3    148.924     681      3   36         11 ...       No       0   
4     55.882     357      2   68         16 ...       No       1   
..       ...     ...    ...  ...        ... ...      ...     ...   
395   12.096     307      3   32         13 ...       No       1   
396   13.364     296      5   65         17 ...       No       0   
397   57.872     321      5   67         12 ...       No       1   
398   37.728     192      1   44         13 ...       No       1   
399   18.701     415      5   64          7 ...       No       0   

            Ethnicity Balance  Sex  
0           Caucasian     333    1  
1               Asian     903    0  
2               Asian     580    1  
3               Asian     964    0  
4           Caucasian     331    1  
..                ...     ...  ...  
395         Caucasian     560    1  
396  African American     480    1  
397         Caucasian     138    0  
398         Caucasian       0    1  
399             Asian     966    0  

[400 rows x 11 columns]

In [129]:
pd.get_dummies(df, columns =['Ethnicity'], drop_first=True)

Income  Rating  Cards  Age  Education         ...          Married  \
0     14.891     283      2   34         11         ...                1   
1    106.025     483      3   82         15         ...                1   
2    104.593     514      4   71         11         ...                0   
3    148.924     681      3   36         11         ...                0   
4     55.882     357      2   68         16         ...                1   
..       ...     ...    ...  ...        ...         ...              ...   
395   12.096     307      3   32         13         ...                1   
396   13.364     296      5   65         17         ...                0   
397   57.872     321      5   67         12         ...                1   
398   37.728     192      1   44         13         ...                1   
399   18.701     415      5   64          7         ...                0   

    Balance  Sex  Ethnicity_Asian  Ethnicity_Caucasian  
0       333    1                0                    1  
1       903    0                1                    0  
2       580    1                1                    0  
3       964    0                1                    0  
4       331    1                0                    1  
..      ...  ...              ...                  ...  
395     560    1                0                    1  
396     480    1                0                    0  
397     138    0                0                    1  
398       0    1                0                    1  
399     966    0                1                    0  

[400 rows x 12 columns]

In [140]:
df.head()

Income  Rating  Cards  Age  Education ...  Student Married  Ethnicity  \
0   14.891     283      2   34         11 ...       No       1  Caucasian   
1  106.025     483      3   82         15 ...      Yes       1      Asian   
2  104.593     514      4   71         11 ...       No       0      Asian   
3  148.924     681      3   36         11 ...       No       0      Asian   
4   55.882     357      2   68         16 ...       No       1  Caucasian   

  Balance  Sex  
0     333    1  
1     903    0  
2     580    1  
3     964    0  
4     331    1  

[5 rows x 11 columns]

In [141]:
pd.get_dummies(df, columns =['Ethnicity'], drop_first=True)
#this is easier, I think

Income  Rating  Cards  Age  Education         ...          Married  \
0     14.891     283      2   34         11         ...                1   
1    106.025     483      3   82         15         ...                1   
2    104.593     514      4   71         11         ...                0   
3    148.924     681      3   36         11         ...                0   
4     55.882     357      2   68         16         ...                1   
..       ...     ...    ...  ...        ...         ...              ...   
395   12.096     307      3   32         13         ...                1   
396   13.364     296      5   65         17         ...                0   
397   57.872     321      5   67         12         ...                1   
398   37.728     192      1   44         13         ...                1   
399   18.701     415      5   64          7         ...                0   

    Balance  Sex  Ethnicity_Asian  Ethnicity_Caucasian  
0       333    1                0                    1  
1       903    0                1                    0  
2       580    1                1                    0  
3       964    0                1                    0  
4       331    1                0                    1  
..      ...  ...              ...                  ...  
395     560    1                0                    1  
396     480    1                0                    0  
397     138    0                0                    1  
398       0    1                0                    1  
399     966    0                1                    0  

[400 rows x 12 columns]

In [93]:
df = df.drop('Ethnicity_Asian', axis=1)

In [142]:
df

Income  Rating  Cards  Age  Education ...  Student Married  \
0     14.891     283      2   34         11 ...       No       1   
1    106.025     483      3   82         15 ...      Yes       1   
2    104.593     514      4   71         11 ...       No       0   
3    148.924     681      3   36         11 ...       No       0   
4     55.882     357      2   68         16 ...       No       1   
..       ...     ...    ...  ...        ... ...      ...     ...   
395   12.096     307      3   32         13 ...       No       1   
396   13.364     296      5   65         17 ...       No       0   
397   57.872     321      5   67         12 ...       No       1   
398   37.728     192      1   44         13 ...       No       1   
399   18.701     415      5   64          7 ...       No       0   

            Ethnicity Balance  Sex  
0           Caucasian     333    1  
1               Asian     903    0  
2               Asian     580    1  
3               Asian     964    0  
4           Caucasian     331    1  
..                ...     ...  ...  
395         Caucasian     560    1  
396  African American     480    1  
397         Caucasian     138    0  
398         Caucasian       0    1  
399             Asian     966    0  

[400 rows x 11 columns]

In [44]:
df = pd.get_dummies(df, columns =['Ethnicity'], drop_first=True)

In [114]:
df

Income  Rating  Cards  Age  Education ...  Student Married  \
0     14.891     283      2   34         11 ...       No       1   
1    106.025     483      3   82         15 ...      Yes       1   
2    104.593     514      4   71         11 ...       No       0   
3    148.924     681      3   36         11 ...       No       0   
4     55.882     357      2   68         16 ...       No       1   
..       ...     ...    ...  ...        ... ...      ...     ...   
395   12.096     307      3   32         13 ...       No       1   
396   13.364     296      5   65         17 ...       No       0   
397   57.872     321      5   67         12 ...       No       1   
398   37.728     192      1   44         13 ...       No       1   
399   18.701     415      5   64          7 ...       No       0   

            Ethnicity Balance  Sex  
0           Caucasian     333    1  
1               Asian     903    0  
2               Asian     580    1  
3               Asian     964    0  
4           Caucasian     331    1  
..                ...     ...  ...  
395         Caucasian     560    1  
396  African American     480    1  
397         Caucasian     138    0  
398         Caucasian       0    1  
399             Asian     966    0  

[400 rows x 11 columns]

> ## Question 3.  Using _sklearn_'s linear regression, predict `Balance` using `Income`, `Cards`, `Age`, `Education`, `Gender`, and `Ethnicity`

In [51]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

/Users/mlombardo/anaconda/lib/python2.7/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


First, find the coefficients of your regression line.

In [144]:
smf.ols(formula = 'Balance ~ Income + Cards + Education + Sex + Ethnicity', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Balance   R-squared:                       0.225
Model:                            OLS   Adj. R-squared:                  0.213
Method:                 Least Squares   F-statistic:                     19.01
Date:                Thu, 27 Jul 2017   Prob (F-statistic):           1.84e-19
Time:                        21:01:03   Log-Likelihood:                -2968.4
No. Observations:                 400   AIC:                             5951.
Df Residuals:                     393   BIC:                             5979.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                138.4452    114.761      1.206      0.228     -87.176     364.067
Ethnicity[T.Asian]         1.1823     57.593      0.021      0.984    -112.048     114.412
Ethnicity[T.Caucasian]     7.1237     50.211      0.142      0.887     -91.592     105.839
Income                     6.0812      0.580     10.482      0.000       4.941       7.222
Cards                     32.2481     14.916      2.162      0.031       2.923      61.573
Education                  1.4955      6.550      0.228      0.820     -11.381      14.372
Sex                      -26.4266     40.836     -0.647      0.518    -106.711      53.858
==============================================================================
Omnibus:                       37.264   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               19.029
Skew:                           0.361   Prob(JB):                     7.38e-05
Kurtosis:                       2.212   Cond. No.                         338.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Then, find the p-values of your F-values' models.  You have a few variables, so try to show your p-values alongside the names of the variables.  (https://docs.python.org/2/library/functions.html#zip)

Asian:     .984
Caucasian: .887
Income:    .000
Cards:     .031
Education: .820
Sex:       .518

> ## Question 4.  Which of your coefficients are significant at the 5% significance level?

Answer: Cards and income

> ## Question 5.  What is your model's $R^2$?

Answer: .225

> ## Question 6.  How do we interpret this value?

Answer: The model explains about 22% of variability of the response data around its mean.

> ## Question 7.  Now let's focus on the two most significant variables from your previous model and re-run your regression model.

In [147]:
smf.ols(formula = 'Balance ~ Income + Cards', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Balance   R-squared:                       0.224
Model:                            OLS   Adj. R-squared:                  0.220
Method:                 Least Squares   F-statistic:                     57.30
Date:                Thu, 27 Jul 2017   Prob (F-statistic):           1.38e-22
Time:                        21:06:10   Log-Likelihood:                -2968.6
No. Observations:                 400   AIC:                             5943.
Df Residuals:                     397   BIC:                             5955.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    151.3299     55.289      2.737      0.006      42.634     260.026
Income         6.0710      0.577     10.525      0.000       4.937       7.205
Cards         31.8381     14.826      2.147      0.032       2.691      60.985
==============================================================================
Omnibus:                       39.698   Durbin-Watson:                   1.975
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               19.444
Skew:                           0.359   Prob(JB):                     5.99e-05
Kurtosis:                       2.194   Cond. No.                         160.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

> ## Question 8.  In comparison to the previous model, did the $R^2$ increase or decrease?  Why?

In [158]:
df

Income  Rating  Cards  Age  Education ...  Student Married  \
0     14.891     283      2   34         11 ...       No       1   
1    106.025     483      3   82         15 ...      Yes       1   
2    104.593     514      4   71         11 ...       No       0   
3    148.924     681      3   36         11 ...       No       0   
4     55.882     357      2   68         16 ...       No       1   
..       ...     ...    ...  ...        ... ...      ...     ...   
395   12.096     307      3   32         13 ...       No       1   
396   13.364     296      5   65         17 ...       No       0   
397   57.872     321      5   67         12 ...       No       1   
398   37.728     192      1   44         13 ...       No       1   
399   18.701     415      5   64          7 ...       No       0   

            Ethnicity Balance  Sex  
0           Caucasian     333    1  
1               Asian     903    0  
2               Asian     580    1  
3               Asian     964    0  
4           Caucasian     331    1  
..                ...     ...  ...  
395         Caucasian     560    1  
396  African American     480    1  
397         Caucasian     138    0  
398         Caucasian       0    1  
399             Asian     966    0  

[400 rows x 11 columns]

Answer: It stayed the same. Honestly, I forget.

> ## Question 9.  Now let's regress `Balance` on `Gender` alone.  After running your linear regressions, do you have enough evidence to claim that females have more balance than males?  (Hint: Look at the p-value of the Gender coefficient.  If it is significant then you will have evidence to support that claim, otherwise you cannot support the statement.)

In [159]:
smf.ols(formula = 'Balance ~ Sex', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Balance   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1836
Date:                Thu, 27 Jul 2017   Prob (F-statistic):              0.669
Time:                        21:12:52   Log-Likelihood:                -3019.3
No. Observations:                 400   AIC:                             6043.
Df Residuals:                     398   BIC:                             6051.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    529.5362     31.988     16.554      0.000     466.649     592.423
Sex          -19.7331     46.051     -0.429      0.669    -110.267      70.801
==============================================================================
Omnibus:                       28.438   Durbin-Watson:                   1.940
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.346
Skew:                           0.583   Prob(JB):                     1.15e-06
Kurtosis:                       2.471   Cond. No.                         2.58
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Answer: I don't think I can support this.

> ## Question 10.  Now let's regress `Balance` on `Ethnicity`.  After running your linear regressions, do you have enough evidence to claim that some ethnic groups carry more balance than others?

In [160]:
smf.ols(formula = 'Balance ~ Ethnicity', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Balance   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.005
Method:                 Least Squares   F-statistic:                   0.04344
Date:                Thu, 27 Jul 2017   Prob (F-statistic):              0.957
Time:                        21:14:23   Log-Likelihood:                -3019.3
No. Observations:                 400   AIC:                             6045.
Df Residuals:                     397   BIC:                             6057.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                531.0000     46.319     11.464      0.000     439.939     622.061
Ethnicity[T.Asian]       -18.6863     65.021     -0.287      0.774    -146.515     109.142
Ethnicity[T.Caucasian]   -12.5025     56.681     -0.221      0.826    -123.935      98.930
==============================================================================
Omnibus:                       28.829   Durbin-Watson:                   1.946
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.395
Skew:                           0.581   Prob(JB):                     1.13e-06
Kurtosis:                       2.460   Cond. No.                         4.39
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Answer: No. P-values are too high.

> ## Question 11.  Finally let's regress `Balance` on `Student`.  After running your linear regressions, do you have enough evidence to claim that students carry more balance than non-students?

In [161]:
smf.ols(formula = 'Balance ~ Student', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Balance   R-squared:                       0.067
Model:                            OLS   Adj. R-squared:                  0.065
Method:                 Least Squares   F-statistic:                     28.62
Date:                Thu, 27 Jul 2017   Prob (F-statistic):           1.49e-07
Time:                        21:14:51   Log-Likelihood:                -3005.5
No. Observations:                 400   AIC:                             6015.
Df Residuals:                     398   BIC:                             6023.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        480.3694     23.434     20.499      0.000     434.300     526.439
Student[T.Yes]   396.4556     74.104      5.350      0.000     250.771     542.140
==============================================================================
Omnibus:                       20.866   Durbin-Watson:                   1.950
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               21.920
Skew:                           0.544   Prob(JB):                     1.74e-05
Kurtosis:                       2.637   Cond. No.                         3.37
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Answer: Yes. Students carry ~$400 more balance than non-students.

> ## Question 12.  No let's consider the effect of `Student` and `Income` on `Balance` simultaneously.  Are all the coefficients significant?

In [164]:
smf.ols(formula = 'Balance ~ Student + Income', data = df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                Balance   R-squared:                       0.277
Model:                            OLS   Adj. R-squared:                  0.274
Method:                 Least Squares   F-statistic:                     76.22
Date:                Thu, 27 Jul 2017   Prob (F-statistic):           9.64e-29
Time:                        21:16:37   Log-Likelihood:                -2954.4
No. Observations:                 400   AIC:                             5915.
Df Residuals:                     397   BIC:                             5927.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        211.1430     32.457      6.505      0.000     147.333     274.952
Student[T.Yes]   382.6705     65.311      5.859      0.000     254.272     511.069
Income             5.9843      0.557     10.751      0.000       4.890       7.079
==============================================================================
Omnibus:                      119.719   Durbin-Watson:                   1.951
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               23.617
Skew:                           0.252   Prob(JB):                     7.44e-06
Kurtosis:                       1.922   Cond. No.                         192.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Answer: Yes, some students are rich and have higer balances?

> ## Question 13.  No let's consider the interaction effect of `Student` and `Income` on `Balance` simultaneously.  Are all the coefficients significant?  It they are, write down your regression model below

(First generate a new variable for the interaction term)

In [ ]:
# TODO

Answer: TODO

> ## Question 14.  Is there any income level at which students and non-students on average carry same level of balance?

Answer: TODO